src/Lumina/Data/LuminaBinaryReader.cs

```diff
@@ -0,0 +1,309 @@
+using System.IO;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Lumina.Data
+{
+    public interface IConvertEndianness
+    {
+        public void ConvertEndianness();
+    }
+
+    public class LuminaBinaryReader : BinaryReader
+    {
+        public PlatformId PlatformId { get; internal set; }
+
+        private bool isLittleEndian;
+        public bool IsLittleEndian
+        {
+            get => isLittleEndian;
+            set => isLittleEndian = value;
+        }
+
+        public bool ConvertEndianness => BitConverter.IsLittleEndian != isLittleEndian;
+
+        public long Position
+        {
+            get => BaseStream.Position;
+            set => BaseStream.Position = value;
+        }
+
+        public LuminaBinaryReader( byte[] array, PlatformId platformId = PlatformId.Win32 ) : this( new MemoryStream( array, false ), Encoding.UTF8, false, platformId )
+        {
+        }
+
+        public LuminaBinaryReader( Stream stream, PlatformId platformId = PlatformId.Win32 ) : this( stream, Encoding.UTF8, false, platformId )
+        {
+        }
+
+        public LuminaBinaryReader( Stream stream, Encoding encoding, PlatformId platformId ) : this( stream, encoding, false, platformId )
+        {
+        }
+
+        public LuminaBinaryReader( Stream stream, Encoding encoding, bool leaveOpen, PlatformId platformId ) : base( stream, encoding, leaveOpen )
+        {
+            PlatformId = platformId;
+            isLittleEndian = platformId != PlatformId.PS3;
+        }
+
+        public override short ReadInt16() => isLittleEndian ? base.ReadInt16() : BinaryPrimitives.ReverseEndianness( base.ReadInt16() );
+        public override ushort ReadUInt16() => isLittleEndian ? base.ReadUInt16() : BinaryPrimitives.ReverseEndianness( base.ReadUInt16() );
+        public override int ReadInt32() => isLittleEndian ? base.ReadInt32() : BinaryPrimitives.ReverseEndianness( base.ReadInt32() );
+        public override uint ReadUInt32() => isLittleEndian ? base.ReadUInt32() : BinaryPrimitives.ReverseEndianness( base.ReadUInt32() );
+        public override long ReadInt64() => isLittleEndian ? base.ReadInt64() : BinaryPrimitives.ReverseEndianness( base.ReadInt64() );
+        public override ulong ReadUInt64() => isLittleEndian ? base.ReadUInt64() : BinaryPrimitives.ReverseEndianness( base.ReadUInt64() );
+        public override float ReadSingle() => isLittleEndian ? base.ReadSingle() : ReverseEndianness( base.ReadSingle() );
+        public override double ReadDouble() => isLittleEndian ? base.ReadDouble() : ReverseEndianness( base.ReadDouble() );
+
+        public unsafe Half ReadHalf()
+        {
+            var half = ReadUInt16();
+
+            return *(Half*)&half;
+        }
+
+        public sbyte[] ReadSBytes( int count ) => MemoryMarshal.Cast< byte, sbyte >( ReadBytes( count ) ).ToArray();
```

Comment:
also these would be far better using generics instead of re-implementing the same thing